torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,255 @@
+
+  bool isNonZeroDimEqual(Tensor& tensor_a, Tensor& tensor_b) {
+    if (tensor_a.dim() != tensor_b.dim()) {
+      return false;
+    }
+    for (int64_t i = 1; i < tensor_a.dim(); i++) {
+      if (tensor_a.size(i) != tensor_b.size(i)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  // Check the linear_layer_group of a tensor to find ones that can be
+  // combined
+  void collectAndMergeLinearLayers(std::list<Node*>& linear_layer_group) {
+    while (!linear_layer_group.empty()) {
+      auto it_first = linear_layer_group.begin();
+      Node* base_node = *it_first;
+
+      std::vector<Node*> compatible_layers;
+      compatible_layers.push_back(base_node);
+      // removing the node from future consideration, as we will only move
+      // nodes forward to keep code simple.
+      linear_layer_group.erase(it_first);
+
+      auto base_weight =
+          constant_as<Tensor>(base_node->namedInput("weight")).value();
+      auto base_bias =
+          constant_as<Tensor>(base_node->namedInput("bias")).value();
+
+      // Now iterate over the rest of the users of the set to
+      // see if there is anything that we can coaleasce `base_node` with.
+      for (auto it = linear_layer_group.begin();
+           it != linear_layer_group.end();) {
+        auto node = *it;
+        auto weight = constant_as<Tensor>(node->namedInput("weight")).value();
+        auto bias = constant_as<Tensor>(node->namedInput("bias")).value();
+
+        // For now we will just keep it simple and require matching types
+        // Type promotion mgiht cause performance to actually decrease.
+        if (base_weight.dtype() != weight.dtype() ||
+            base_weight.device() != weight.device() ||
+            base_bias.dtype() != bias.dtype() ||
+            base_bias.device() != bias.device()) {
+          it++;
+          continue;
+        }
+
+        if (!isNonZeroDimEqual(base_weight, weight) ||
+            !isNonZeroDimEqual(base_bias, bias)) {
+          it++;
+          continue;
+        }
+
+        bool can_move_before_all = true;
+        for (auto n : compatible_layers) {
+          can_move_before_all &=
+              getAliasDb()->couldMoveBeforeTopologically(node, n);
+        }
+        if (!can_move_before_all) {
+          it++;
+          continue;
+        }
+
+        // Found a node that is eligible for combination
+        auto prev = it;
+        it++;
+        linear_layer_group.erase(prev);
+        compatible_layers.push_back(node);
+      }
+      if (compatible_layers.size() == 1) {
+        continue; // No other layers to merge
+      }
+      mergeLinearLayers(base_node, compatible_layers);
+    }
+  }
+
+  void handleBlockAndSubblocks(Block* block) {
+    for (auto node : block->nodes()) {
+      for (Block* block : node->blocks()) {
+        handleBlockAndSubblocks(block);
+      }
+    }
+
+    // Processing for the block itself
+    std::unordered_map<Value*, std::list<Node*>> grouped_linear_layers;
```

Comment:
So, the way that this would be a reverse topological iteration through the graph is if we did
```
block0:
     node1
     node2
           block1 
     node3
           block2
                  block3
```
block3->block2->node3->block1->node2->node1. However, what we are doing instead is block3->block2->block1->node3->node2->node1. I think we should be fine since we dont descend into block1 in alias analysis when we are processing node1/node2.... I think we are fine here since we are only looking at linear nodes in the same block and wouldnt look at any nodes in block1 while we are iterating block0, and we don't descend into blocks when trying the tryMoveTopologicallyValid as of now... just something to think about, we might want to add a test to make sure that if  alias analysis tryMoveTopologicallyValid did change we dont get an error. 


Something like :
```
def foo(x, cond: bool):
     b = c = torch.zeros([20, 20])
     bias = torch.zeros([20])
     out3 = torch.nn.functional.linear(x, b, bias)
     if cond:
          out1 = torch.nn.functional.linear(x, b, bias)
          out2 = torch.nn.functional.lienar(x, c, bias)
    else:
          assert False
    out4 = torch.nn.functional.linear(x, c, bias)
    return out1, out2, out3, out4
```
 
This way we check that trying to merge together out4 and out3 doesnt yield an error